PARCEL_2_RFC.md

```diff
@@ -0,0 +1,1285 @@
+```sh
+$ parcel [...entries] [...flags]
+```
+
+#### `[...entries]`
+
+Entry files to start bundling, these will be preserved as entry points in the
+output. Defaults to `package.json#source`, falling back to `src/index.*` or
+`index.*`. See [#Entries](#entries-sources-targets-environment).
+
+#### `--serve, -s`
+
+Serve assets on a local server
+
+#### `--watch, -w`
+
+Watch and rebuild code on file changes.
+
+#### `--open, -o [browser]`
+
+Open your local server in a browser. You can optionally pass the name of the
+browser you want to open, otherwise it will use your default browser.
+
+#### `--port <port>`
+
+Configure the port to serve assets on. Alternatively you can use the `$PORT`
+environment variable.
+
+#### `--https`
+
+This will generate a local certificate (which will be untrusted by your
+browser, you'll need to approve it) and serve your assets over `https://`
+
+##### `--cert <path>`
+
+Specify the filepath to your SSL certificate when using `--https`.
+
+##### `--key <path>`
+
+Specify the filepath to your SSL key when using `--https`.
+
+#### `--cache <dir>`, `--no-cache`
+
+Configure the cache directory with `--cache <dir>` or disable it altogether
+with `--no-cache`.
+
+#### `--hot`, `--no-hot`
+
+Turn hot reloading on or off.
+
+##### `--hot-hostname <hostname>`
+
+Configure the hot reloading hostname.
+
+##### `--hot-port <port>`
+
+Configure the hot reloading port.
+
+#### `--[no-]source-maps`
+
+Turn source maps on or off. Source maps are turned on by default except in
+production builds.
+
+#### `--autoinstall [npm/yarn], --no-autoinstall`
+
+When enabled, whenever Parcel discovers a dependency that isn't installed it
+will attempt to install it with either npm or Yarn (defaults to npm unless a
+`yarn.lock` exists).
+
+#### `--mode <mode>`
+
+Override the environment mode, including those manually configured to something
+else (ex. `--mode production` or `--mode development`). Defaults to the
+`NODE_ENV` environment variable.
+
+##### `--development, --dev`
+
+Aliases for `--mode development`.
+
+##### `--production, --prod`
+
+Aliases for `--mode production`.
+
+##### `--test`
+
+Aliases for `--mode test`.
+
+#### `--public-url <url>`
+
+[todo]
+
+#### `--log-level <level>`
+
+Set the log level, either "0" (no output), "1" (errors), "2" (warnings +
+errors) or "3" (all). (Default: 2)
+
+#### `--version, -v, -V`
+
+Return the current version of Parcel.
+
+#### `--help, -h`
+
+Get help with the CLI.
+
+## Parcel Config
+
+Parcel has always and will always work out of the box for many projects with
+zero configuration. It should always be extremely simple to get started. But if
+you do want more control, we give you the tools to do so.
+
+### Configuring external tools
+
+A huge part of what Parcel does is run other tools over your code. Instead of
+pulling all that configuration into Parcel, we make use of their own
+configuration systems. So if you're using Babel, you should just use `.babelrc`
+files to configure it.
```

Comment:
What if the tool does not support dynamic configuration files (that for example depend on the filepath that's being processed, and has a custom logic like substituting a parser based on some condition)? Parcel still uses the tools via their JS APIs, so it could potentially resolve and pass a generated configuration to them. I keep webpack mainly because it can do that and that I can generate a different babelrc for different parts of the app (node processes) within a single package, but Parcel requires the root of your package to be occupied by various tools' config files that are the same for the entire code base. Of course, in the ideal world, different parts would be separate packages, reusable parts in their own packages, etc., but it's a rare luxury to have a large real world app code structured like that. What do you think?

Reply:
how about a configuration-check to test if the `.parcelrc` is a JSON format or otherwise try to evaluate it as JS file to dynamically parse the configuration, like in webpack?

example:

### JSON format
```
{
  "extends": ["@parcel/config-default"],
  "resolvers": ["@parcel/resolver-default", "./config/resolvers"],
  "bundler": "@parcel/bundler-default",
  [...]
}
```

### JS format
```
const myCustomResolverConfig = require("./config/resolvers");

module.exports = {
  "extends": ["@parcel/config-default"],
  "resolvers": ["@parcel/resolver-default", ...myCustomResolverConfig ],
  "bundler": "@parcel/bundler-default",
  [...]
}
```

if we do like that and parcel detects a JS-Config, it should default to parse config and override tool-specific configuration. then the tool itself gets the original config, overriden by parcel with the override-values of it.

so e.g. `.babelrc` contains a `transforms`-field but we explicitly override it in `.parcelrc`, then the final configuration would be the diff of both, where `.parcelrc` takes priority.

Reply:
We will not be supporting a dynamic config format in JS or any other turing complete language. If tools do not support being configured the way you want, you'll have to bring that up with them.

Reply:
Let's hope tool authors will get on the parcel train and start introducing better APIs when parcel v2 comes out.
But maybe it introduces new tools which target such scenarios when using parcel and that's what I'm hoping for personally.